### PR TITLE
Move dist implementation into chef-utils

### PR DESCRIPTION
**Pull request by bobchaos**
Signed-off-by: Marc Chamberland <chamberland.marc@gmail.com>

## Description
All dist contants are moving to chef-utils under new namespaces for each product, ie: `ChefUtils::Dist::Infra::PRODUCT`. 

~~All that remains to be done is to go through all the relevant files' `require` statement to elemininate duplicate `require "chef-utils"` statement generated by automation and require only dist where other chef-utils features aren't in use.~~ Being enormous, I figured I'd open the PR sooner than later. Sorry :/

## Related Issue
As discussed on slack with @lamont-granquist 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [N/A] I have updated the documentation accordingly. - This new implementation may be worth documenting, lemme know if you think so and I'll find some place for it in the docs
- [N/A] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by bobchaos** [comment]
This should be ready to process but I'd like to hold off as long as possible on merging it so I can write a blog post or something informing redistributors that we're about to break everything o.O

**Comment by bobchaos** [comment]
I'm seeing test failures here, but no recent PR is green... Should I worry? FWIW I looked through them and found nothing that screams "This is because of the dist implementation", and I have no error that are not also present in a PR dated a few hours before this one.

File: lib/chef/application.rb
Comment (tas50) [inline lib/chef/application.rb:30]: Let's require the specific part of chef-utils we need vs. the whole gem

File: chef-utils/lib/chef-utils/dist/org.rb
Comment (tas50) [inline chef-utils/lib/chef-utils/dist/org.rb:1]: Can we keep this class structure, but stick it all into a singel file so it's easier to read and require?

File: chef-utils/lib/chef-utils.rb
Comment (tas50) [inline chef-utils/lib/chef-utils.rb:32]: We can skip this part and continue to require dist when we need it within the specific ruby files

File: lib/chef/application.rb
Comment (bobchaos) [reply lib/chef/application.rb:30]: where I put a plain `require "chef-utils"` as opposed to `chef-utils/dist` it's because there was another chef-utils function being used, in this case `.windows?`. I could make 2, more specific require statements tho, would that be better?

File: chef-utils/lib/chef-utils/dist/org.rb
Comment (bobchaos) [reply chef-utils/lib/chef-utils/dist/org.rb:1]: arg, I had it like that at one point too XD I'll fix it, should be much easier now that I've done an initial pass.

**Comment by tas50** [comment]
@bobchaos can you rebase this. Merging #9878 with the dist fix broke this.

**Comment by ramereth** [comment]
@tas50 @lamont-granquist I'm working on rebasing and updating this PR and was wondering if we should provide backwards compatibility support for ``lib/chef/dist.rb`` so give us a migration path for downstream gems. We could replace the references in that file to use their ``ChefUtils::Dist`` counter parts. The current version of this PR completely removes ``lib/chef/dist.rb``. Not sure if we should add some kind of deprecation warning or not. I figure once this gets merged and released, we can work and fix any downstream gems that depend on this.

Also, should all of the references in ``chef-config/lib/chef-config/dist.rb`` be migrated to ``chef-utils/lib/chef-utils/dist.rb`` as well? And the same question above regarding a migration path.

Please let me know what you think.

**Comment by jgitlin-p21** [comment]
My personal vote would be to deprecate `lib/chef/dist.rb` and `chef-config/lib/chef-config/dist.rb` by keeping the files but adding a deprecation notice, and making their values equal to the `ChefUtils::Dist` counter parts, with the `lib/chef/dist.rb` and `chef-config/lib/chef-config/dist.rb` files to be removed completely in a future version.

Just my $0.02

**Comment by ramereth** [comment]
FWIW I couldn't find any direct references to ``Chef::Dist`` or ``ChefConfig::Dist`` outside of this repository, so it might be safe to completely remove/deprecate both of those.

I did some additional investigation of all the places where ``dist.rb`` exists and found the following references:
- https://github.com/chef/omnitruck/blob/ab4ca84865decd69d2c41a609bc9e3bf43bc5019/lib/dist.rb
- https://github.com/chef/mixlib-install/blob/13bbdce54ef9e48a0cf25bccc4152bf76abbfaaa/lib/mixlib/install/dist.rb
- https://github.com/chef/chef-cli/blob/f521a80e3be037eede70cf7734326d82f4b5d881/lib/chef-cli/dist.rb
- https://github.com/chef/chef-zero/blob/ed38f23fef793f73fb690e5ea59bf32b57a994c5/lib/chef_zero/dist.rb
- https://github.com/chef/chef-dk/blob/67c02e3a500efaaef5337caed503296fd443ede7/lib/chef-dk/dist.rb
- https://github.com/inspec/inspec/blob/a35d37043e65ce5fd528d58cd48ecb369c74a432/lib/inspec/dist.rb

I know that we're replacing chef_zero and inspec already in this PR, but I'm wondering if all of those should eventually be migrated here as well(well, except for chef-dk since that's being deprecated). I'd be curious what others think. That would create including chef-utils as a gem dependency in each of those projects.

**Comment by jgitlin-p21** [comment]
> That would create including chef-utils as a gem dependency in each of those projects

From these comments in the `#chef-infra-dev` slack channel:

> **tas50:** chef-config is parsing out the config files.
> chef-utils is various helpers that can be used throughout our products.
> the OS detection stuff is all in there
> which / shell_out helpers are in there
> it's a dep free helper collection
> **lamont:** chef-utils is designed to not have any dependencies on anything else to avoid circular deps and to be broadly usable across literally everything

It seems to me like the answer is "Yes, a dependency is fine and `chef-utils` is the proper place so all of those should eventually be migrated here as well" but again just adding my $0.02 😆

**Comment by lamont-granquist** [comment]
Yes, we should kill off the exiting locations.  I think/hope that nobody has deps on those, definitely I wouldn't consider that a public API yet, we're still figuring it all out.  And yes, we need to have all the other dist.rb's removed and those projects need to take a dep on chef-utils and use the dist constants out of chef-utils.

That is mildly terrifying, and would suggest we do this in a few phases:

1. omnitruck/mixlib-install/chef-zero -- these are obviously relatively "safe"
2. chef-cli -- this has tighter integration with chef-client already, but is "bigger"
3. inspec -- this will be a whole new world

You can probably omit chef-dk since it is basically dead now.

File: lib/chef/application.rb
Comment (ramereth) [reply lib/chef/application.rb:30]: I just realized I may have reverted his change when I rebased this. Do you want me to split this up to requiring ``chef-utils/dist`` and ``chef-utils/dsl/windows`` like it was before?

File: lib/chef/application/client.rb
Comment (ramereth) [inline lib/chef/application/client.rb:24]: Should I change this to the following?
```suggestion
require "chef-utils/dsl/windows"
require "chef-utils/dist"
```

File: lib/chef/application/solo.rb
Comment (ramereth) [inline lib/chef/application/solo.rb:24]: Should I change this to the following?
```suggestion
require "chef-utils/dsl/windows"
require "chef-utils/dist"
```

File: lib/chef/client.rb
Comment (ramereth) [inline lib/chef/client.rb:58]: Looks like maybe I should have this changed to the following?
```ruby
require "chef-utils/dist"
```

File: lib/chef/event_loggers/windows_eventlog.rb
Comment (ramereth) [inline lib/chef/event_loggers/windows_eventlog.rb:22]: Should this be?
```suggestion
require "chef-utils/dsl/windows"
require "chef-utils/dist"
```

File: lib/chef/knife.rb
Comment (ramereth) [inline lib/chef/knife.rb:35]: This should probably be the following, correct?
```suggestion
require "chef-utils/dist"
```

File: lib/chef/local_mode.rb
Comment (ramereth) [inline lib/chef/local_mode.rb:18]: Should this be?
```suggestion
require "chef-utils/dist"
```

File: lib/chef/provider/file.rb
Comment (ramereth) [inline lib/chef/provider/file.rb:34]: Should this be?
```suggestion
require "chef-utils/dist"
```

File: spec/functional/event_loggers/windows_eventlog_spec.rb
Comment (ramereth) [inline spec/functional/event_loggers/windows_eventlog_spec.rb:22]: ?
```suggestion
require "chef-utils/dist"
```

File: spec/integration/solo/solo_spec.rb
Comment (ramereth) [inline spec/integration/solo/solo_spec.rb:8]: ?
```suggestion
require "chef-utils/dist"
```

File: lib/chef/application/client.rb
Comment (lamont-granquist) [reply lib/chef/application/client.rb:24]: nope

File: lib/chef/client.rb
Comment (lamont-granquist) [reply lib/chef/client.rb:58]: yeah you should need to change that, that file shouldn't exist so something should be blowing up in tests?

Comment (ramereth) [reply lib/chef/client.rb:58]: Actually this is covered with via the require on ``chef-utils`` above so this can be ignored.

File: lib/chef/application.rb
Comment (ramereth) [reply lib/chef/application.rb:30]: I think based on the conversation I had with @lamont-granquist this morning, this is probably better. to leave as it.

**Comment by ramereth** [comment]
@tas50 @lamont-granquist I believe this is ready for another review. I'm not sure if the failing tests are related or not.

**Comment by phiggins** [comment]
> I'm not sure if the failing tests are related or not.

The failing unit tests definitely look related:

```
  1) ChefConfig::Config#var_chef_path when forced to unix etc_chef_dir is /etc/chef
     Failure/Error: let (:dirname) { ChefConfig::Dist::DIR_SUFFIX }
 
     NameError:
       uninitialized constant ChefConfig::Dist
       Did you mean?  Digest
     # ./spec/unit/config_spec.rb:228:in `block (3 levels) in <top (required)>'
     # ./spec/unit/config_spec.rb:268:in `block (4 levels) in <top (required)>'
```

https://buildkite.com/chef-oss/chef-chef-master-verify/builds/6050#65521c8e-b41b-4803-906b-73ddaf9b14f3/88-1794

File: Rakefile
Comment (tas50) [inline Rakefile:29]: you might want to slack the defined? bit in here to avoid that chefstyle issue when we bump that in chef/chef

**Review by lamont-granquist — APPROVED** [review]

**Pull request merged**
